src/preprocessing/passes/quantifier_macro_elimination.cpp

```diff
@@ -0,0 +1,249 @@
+        Trace("QME") << "found definition:\n  " << predicate
+                     << " := " << definition << endl;
+        d_macroDefinitions[predicate] = definition;
+        d_macroBoundVariables[predicate] = application_children;
+        // don't bother rewriting in definition node
+        d_definitionNodes.insert(application);
+      }
+    }
+  }
+}
+
+Node QuantifierMacroElimination::replaceMacroInstances(Node n)
+{
+  Trace("QME:all") << "replaceMacroInstances: " << n << endl;
+  if (setContains(d_definitionNodes, n))
+  {
+    Trace("QME:all") << "skipping: " << n << endl;
+    return n;
+  }
+  // TODO question: more obvious way to get a children iterator?
+  vector<Node> children;
+  for (size_t i = 0; i < n.getNumChildren(); ++i)
+  {
+    children.push_back(n[i]);
+  }
+  if (n.getKind() == APPLY_UF)
+  {
+    Node predicate = n.getOperator();
+    if (d_macroDefinitions.find(predicate) != d_macroDefinitions.end())
+    {
+      Node macroDefinition = d_macroDefinitions[predicate];
+      vector<Node> macro_variables = d_macroBoundVariables[predicate];
+      Trace("QME") << "applying substitution for\n  " << n << endl;
+      auto new_node = macroDefinition.substitute(macro_variables.begin(),
+                                                 macro_variables.end(),
+                                                 children.begin(),
+                                                 children.end());
+      Trace("QME") << "rewritten node: " << new_node << endl;
+      return new_node;
+    }
+    else
+    {
+      return n;
+    }
+  }
+  else
+  {
+    vector<Node> newChildren;
+    if (n.getNumChildren() > 0)
+    {
+      for (size_t i = 0; i < n.getNumChildren(); ++i)
+      {
+        // TODO don't recurse
+        newChildren.push_back(replaceMacroInstances(n[i]));
+      }
+      Assert(newChildren.size() == children.size());
+      if (n.getMetaKind() == metakind::PARAMETERIZED)
+      {
+        newChildren.insert(newChildren.begin(), n.getOperator());
+      }
+      return NodeManager::currentNM()->mkNode(n.getKind(), newChildren);
+    }
+    else
+    {
+      return n;
+    }
+  }
+}
+
+PreprocessingPassResult QuantifierMacroElimination::applyInternal(
+    AssertionPipeline* assertionsToPreprocess)
+{
+  d_preprocContext->spendResource(options::preprocessStep());
+
+  std::vector<Node>& assertions = assertionsToPreprocess->ref();
+  // Find macro definitions:
+  for (Node& assertion : assertions)
+  {
+    addMacroDefinitionIfPresent(assertion);
+  }
+  // TODO remove macro definition assertions from problem?
+  Trace("QME") << "definitions:\n\n" << d_macroDefinitions << endl << endl;
+  Trace("QME") << "definition bindings:\n" << d_macroBoundVariables << endl;
+  // Apply macro substitutions:
+  if (d_macroDefinitions.size() == 0)
+  {
+    return PreprocessingPassResult::NO_CONFLICT;
+  }
+  for (Node& assertion : assertions)
+  {
+    // TODO currently 10 quantifier regression tests fail when check-unsat-cores
```

Comment:
I think Andres pointed me to this being the point where something should be done to resolve this issue but I haven't looked into it yet

Reply:
Yes, you can look at the old version of the macros preprocessing pass to see how to do this:  https://github.com/CVC4/CVC4/blob/master/src/preprocessing/passes/quantifier_macros.cpp#L104